src/rust/engine/src/nodes.rs

```diff
@@ -760,6 +761,41 @@ impl From<RunId> for NodeKey {
   }
 }
 
+#[derive(Clone, Debug, DeepSizeOf, Eq, Hash, PartialEq)]
+pub struct CurrentGoals;
+
+impl CurrentGoals {
+  async fn run_node(self, context: Context, workunit: &mut RunningWorkunit) -> NodeResult<Value> {
+    let gil = Python::acquire_gil();
+    let py = gil.python();
+    let names = workunit.get_current_goals();
```

Comment:
Rather than doing this via workunits, you should probably make it a mutable property of the `Session`, similar to the `SessionValues`: https://github.com/pantsbuild/pants/blob/7f364f8263eb0a2f34a2a6eebe618e04db5da2ab/src/rust/engine/src/session.rs#L86-L87

Then, in [run_goal_rules](https://github.com/pantsbuild/pants/blob/7f364f8263eb0a2f34a2a6eebe618e04db5da2ab/src/python/pants/init/engine_initializer.py#L118-L158), you would try-finally setting and clearing the `CurrentGoals` value.

Reply:
Ahh.. ok. Interesting. Yea, going via workunits has been a source of much headache and fiddling to tie the pieces together.

Reply:
But wait.. if we have that in `try/finally` in the python world, that's pretty close to what I had originally only applied at a different spot, right?

https://github.com/pantsbuild/pants/commit/aaa50d658567a0fa913932e7b430d17bbfbd76be

Where I stored it in `SessionValues` directly rather than as a property of the session, though.

Reply:
Hmm. Yea, it is.

I think that I was confused by the mutation in that case... I think that you could store this directly in the `SessionValues`, which we already support setting?